README.md update (#42046)
qgis/QGIS: README.md

<img src="images/README-md/main_logo.png" width="300">

[![Tests](https://github.com/qgis/QGIS/workflows/QGIS%20tests/badge.svg)](https://github.com/qgis/QGIS/actions?query=workflow%3A%22QGIS+tests%22)
[![Docker Status](https://img.shields.io/docker/automated/qgis/qgis.svg)](https://hub.docker.com/r/qgis/qgis/tags)
[![Build Status](https://dev.azure.com/qgis/QGIS/_apis/build/status/qgis.QGIS?branchName=master)](https://dev.azure.com/qgis/QGIS/_build/latest?definitionId=1&branchName=master)
![Windows cross build with MXE](https://github.com/qgis/QGIS/workflows/Windows%20cross%20build%20with%20MXE/badge.svg)

> # About QGIS

Reviewer: How about using https://freelance-tech-writer.github.io/table-of-contents-generator/index.html to generate a table of contents at the beginning, as done in the install.md file?

Reviewer 2: Good idea. Let's get this pull committed and do this in next round. I'm hoping people come up with even better visual examples to rotate in too.

Reviewer 2: @benhur07b tells me he explored this, but it pushes the features section further down the page. Think our goal is to get that and the visuals in front of eyeballs as close to the top as possible.

Reviewer: I didn't try myself but maybe it depends on how deep we want to expand the toc

QGIS is a full-featured, user-friendly, free-and-open-source (FOSS) geographical information system (GIS) that runs on Unix platforms, Windows, and MacOS.

## Features 

### 1. Flexible and powerful spatial data management

- Supports raster, vector, mesh, and point cloud data in a range of industry-standard formats
    - *Raster formats include:* GeoPackage, GeoTIFF, GRASS, ArcInfo binary and ASCII grids, ERDAS Imagine SDTS, WMS, WCS, PostgreSQL/PostGIS, and [other GDAL supported formats](https://gdal.org/drivers/raster/index.html).
    - *Vector formats include:* GeoPackage, ESRI Shapefiles, GRASS, SpatiaLite, PostgreSQL/PostGIS, MSSQL, Oracle, WFS, Vector Tiles and [other OGR supported formats](http://www.gdal.org/ogr_formats.html). 
    - *Mesh formats include*: NetCDF, GRIB, 2DM, and [other MDAL supported formats](https://github.com/lutraconsulting/MDAL#supported-formats).
    - *Point-cloud format*: LAS/LAZ and EPT datasets.

> - Access and display local files, spatial databases (PostGIS, SpatiaLite, MSSQL, SAP HANA), web services (WMS, WCS, WFS, ArcGIS REST services), tile services, etc. <!-- removed DB2 due to https://github.com/qgis/QGIS/pull/41178 -->

Reviewer 3: ```suggestion
- Access and display local files, spatial databases (PostGIS, SpatiaLite, SQL Server, Oracle, SAP HANA), web services (WMS, WCS, WFS, ArcGIS REST services), tile services, etc. <!-- removed DB2 due to https://github.com/qgis/QGIS/pull/41178 -->
```

- Geospatial database engine (filters, joins, relations, forms, etc.), as close to datasource- and format-independent as possible 
- Immediate visualization of geospatial query and geoprocessing results
- Graphical modeler and batch processing

*Example: Travel isochrones*

![Example: Travel isochrones](images/README-md/network_analysis_2.png "Travel isochrones")

*Example: Graphical model*

![Example: graphical model](https://docs.qgis.org/latest/en/_images/models_model.png "Graphical model")

### 4. Powerful customization and extensibility

- Fully customizable user experience, including user interface and application settings that cater to power-users and beginners alike
- Rich [expression engine](https://docs.qgis.org/testing/en/docs/user_manual/working_with_vector/expression.html) for maximum flexibility in visualization and processing
- Broad and varied [plugin ecosystem](https://plugins.qgis.org/) that includes data connectors, digitizing aids, advanced analysis and charting tools, in-the-field data capture, etc.
- Style manager for creating, storing, and managing styles
- [QGIS style hub](https://plugins.qgis.org/styles/) for easy sharing of styles

> - Python and C++ API for standalone (headless) applications as well as in-application comprehensive scripting (PyQGIS)

Reviewer: Link the pyqgis and c++ api website here, or are they mentioned somewhere else?

Reviewer 2: Good idea will add.

Reviewer 2: Done

*Example: Style manager*

![Example: Style manager](https://docs.qgis.org/latest/en/_images/stylemanager.png "Style Manager")

*Example: Plugins*

![Example: Plugins](images/README-md/plugins_1.png "Plugins")

<!-- Kill this one for now, since it's Python2 not 3
Example: Python console

> ![Example: Python console](https://docs.qgis.org/latest/en/_images/python_console_editor.png "Python console")

Reviewer: we'll be happy to get this fixed in the docs :smile:

Reviewer 2: :) -- though not sure we need to include it here anyway. Think best to limit to 2 examples per section, and python code is python code. We threw this one off the island because of the 2 vs 3, but think we ended up with a better set of 2 complementary ex anyway.

-->

### 5. QGIS server

Headless map server -- running on Linux, macOS, Windows, or in a docker container -- that shares the same code base as QGIS.

- Industry-standard protocols (WMS, WFS, WFS3/OGC API for Features and WCS) allow plug-n-play with any software stack
- Works with any web server (Apache, nginx, etc) or standalone

> - All QGIS beautiful cartography is supported with best-in-class support for printing
> - Fully customizable with Python scripting support

Reviewer 3: ```suggestion
- All beautiful QGIS cartography is supported with best-in-class support for printing
- Fully customizable with Python scripting support
```

*Example: QGIS server WMS response*

> ![Example: QGIS Server response to a WMS request](https://docs.qgis.org/latest/en/_images/server_selection_parameter.png "QGIS Server response to a WMS request")

Reviewer 2: Good point. I'm torn between making everything self-contained in this repo for greater stability, but then it sits and slowly ages (like the Python 2 vs 3 script screenshot in the docs), vs reuse. In the balance, for now we decided to feel free to point to images on our own website, plus the QGIS flickr group which seems stable enough we scrape it at https://www.qgis.org/en/site/about/screenshots.html already. But we pulled other images into ./images/README-md/

Bottom line is if our own links rot, we'll notice pretty quickly and it's easy to fix. And I'd love to get this into master this way to provoke people to come up with even better examples we can then reuse in docs. If we continue to be worried about this, any image can be copied to ./images/README-md/ whenever desired.

> ![Example: QGIS Server response to a WMS request](https://docs.qgis.org/latest/en/_images/server_selection_parameter.png "QGIS Server response to a WMS request")

Reviewer: OK. Let's keep as is and hope we catch things when/if they break.

*Example: QGIS server WFS response*

![Example: QGIS Server response to a WFS Feature request](https://docs.qgis.org/latest/en/_images/server_wfs3_feature.png "QGIS Server response to a WFS Feature request")

## Under the hood

QGIS is developed using the [Qt toolkit](https://qt.io) and C++, since 2002, and has a pleasing, easy to use graphical
user interface with multilingual support. It is maintained by an active developer team and supported by vibrant 
community of GIS professionals and enthusiasts as well as geospatial data publishers and end-users. 

### Versions and release cycle

QGIS development and releases follow a [time based schedule/roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html). There are three main branches of QGIS that users can install. These are the **Long Term Release (LTR)** branch, **Latest Release (LR)** branch, and the **Development (Nightly)** branch. 

> Every month, there is a **Point Release** that provides bug-fixes to the LTR and bug-fixes as well as new features to the LR.

Reviewer 2: Could have philosophical/purist discussion about it, but I think we've been adding enough enhancements into point releases that it's valid as is.

> Every month, there is a **Point Release** that provides bug-fixes to the LTR and bug-fixes as well as new features to the LR.

Reviewer: > I think we've been adding enough enhancements into point releases that it's valid as is.

I'm not sure of this and even then, it's not deliberate (and are they not ported to LTR?). IMO it looks confusing to expose this kind of info to users. Afaiu, new feature = next release. Any released version only gets fixes, either for 3 months (LR) or for 15 (LTR).

> Every month, there is a **Point Release** that provides bug-fixes to the LTR and bug-fixes as well as new features to the LR.

Reviewer 2: OK. I simplified to "provides bug-fixes to the LTR and LR".

## Installing and using QGIS

Precompiled binaries for QGIS are available at [the QGIS.org download page](https://www.qgis.org/en/site/forusers/download.html).
Please follow the installation instructions carefully.

The [building guide](INSTALL.md) can be used to get started with building QGIS from source.

For installation of QGIS Server, see its [getting started documentation](https://docs.qgis.org/testing/en/docs/server_manual/getting_started.html).

### Documentation

A range of 
[documentation](https://qgis.org/en/docs/index.html) is available. This includes

- [Training Manual](https://docs.qgis.org/latest/en/docs/training_manual/index.html)
- [QGIS User Guide](https://docs.qgis.org/latest/en/docs/user_manual/index.html)
- [QGIS Server Guide](https://docs.qgis.org/latest/en/docs/server_manual/index.html)
- [Documentation Guidelines](https://docs.qgis.org/latest/en/docs/documentation_guidelines/index.html)
- [PyQGIS Cookbook](https://docs.qgis.org/latest/en/docs/pyqgis_developer_cookbook/index.html)
- [Developers Guide](https://docs.qgis.org/latest/en/docs/developers_guide/index.html)
- [Visual Changelog](https://qgis.org/en/site/forusers/visualchangelogs.html)

> ### Help and Support Channels

Reviewer 3: ```suggestion
### Help and support channels
```

There are several channels where you can find help and support for QGIS:

> - Using the QGIS community site at [https://qgis.org](https://qgis.org)

Reviewer: ```suggestion
- Using the QGIS community site at https://qgis.org
```

Reviewer 2: Not sure what this change adds? Parallelism with other links isn't bad. But happy to change if important.

Reviewer: Maybe, parallelism could be 
```suggestion
- Using the QGIS community [site](https://qgis.org)
```
Looks weird to me to encapsulate a hyperlink within a hyperlink. Users would anyway read the url in this case so...
bref... no strong opinion.

Reviewer 2: OK. I like that. fixed.

Reviewer 2: Fixed with [QGIS community site](...)

    - Using a Matrix client and joining the [#qgis:matrix.org](http://matrix.to/#/#qgis:matrix.org) room.
    - Using [Gitter](https://gitter.im/qgis/QGIS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) chat.
 - At the [GIS stackexchange](https://gis.stackexchange.com/) or [r/QGIS reddit](https://www.reddit.com/r/QGIS/), which are not maintained by the QGIS team, but where the QGIS and broader GIS community provides lots of advice
- [Other support channels](https://qgis.org/en/site/forusers/support.html)

## Get involved with the community

### Bug reporting and bug fixing

You can help us **by submitting bug reports or fixing bugs** in the [QGIS bug tracker](https://github.com/qgis/QGIS/issues/).

### New features and enhancements

If you wish to contribute patches you can [fork the project](https://help.github.com/forking/), make your changes, commit to your repository, and then [create a pull request](https://help.github.com/articles/creating-a-pull-request-from-a-fork/). The development team can then review your contribution and commit it upstream as appropriate.

> If you commit a new feature, add [FEATURE] to your commit message AND give a clear description of the new feature. A webhook will automatically create an issue on the QGIS-Documentation repo to tell people to write documentation about it.

Reviewer 2: Sorry, not following exactly what change is needed here. I'll invite you as contributor to my branch/fork and you can adjust this (and other things) directly there, if you wish, or post-merge in master of course.

> If you commit a new feature, add [FEATURE] to your commit message AND give a clear description of the new feature. A webhook will automatically create an issue on the QGIS-Documentation repo to tell people to write documentation about it.

Reviewer: I did not make a suggestion because I failed to find the right balance and wondered if it should be removed. Here's a suggestion anyway.
```suggestion
If you commit a new feature, add `[FEATURE]` to your commit message AND give a clear description of the new feature. The label `Needs documentation` will be added by maintainers and will automatically create an issue on the QGIS-Documentation repo to tell people to write documentation about it.
```
But I don't like the "tell people to write documentation" part since it may convey the message that the developer is not welcome to doc his work himself (there's already people to do that).

> If you commit a new feature, add [FEATURE] to your commit message AND give a clear description of the new feature. A webhook will automatically create an issue on the QGIS-Documentation repo to tell people to write documentation about it.

Reviewer 2: Ah, now I understand. I just put in the following, based on your suggestion but inviting documentation by the contributor:

> The label `Needs documentation` will be added by maintainers and will automatically create an issue on the QGIS-Documentation repo, where you or others should write documentation about it.

> If you commit a new feature, add [FEATURE] to your commit message AND give a clear description of the new feature. A webhook will automatically create an issue on the QGIS-Documentation repo to tell people to write documentation about it.

Reviewer: Nicer

For large-scale changes, you can open a [QEP (QGIS Enhancement Proposal)](https://github.com/qgis/QGIS-Enhancement-Proposals). QEPs are used in the process of creating and discussing new enhancements or policy for QGIS. 

### Translations

Please help translate QGIS to your language. At this moment about forty languages are already available in the Desktop user interface and about eighty languages are available in transifex ready to be translated.

The [translation](https://qgis.org/en/site/getinvolved/translate.html) process is managed by the [Translation Team](https://qgis.org/en/site/getinvolved/governance/governance.html#gui-translation) and all the activities are done under the [Transifex](https://www.transifex.com/) platform.

> 

Reviewer: Is it worth to mention that every language that is translated at 35% is automatically pulled in the release? People often wonder why their language is not pulled or not pulled anymore.

Reviewer 2: Interesting. Feels to me like it's going a bit too much into the weeds to put this into the overall README.md, but seems like a good point for the "how to contribute effectively if you're not a developer" landing page to be created (somewhere?) as per comment elsewhere in this discussion.

### Other ways to contribute

> If you are not a developer, there are many other possibilities which do not require programming skills to help QGIS to evolve. Check our [project homepage for more information](http://qgis.org/en/site/getinvolved/index.html).

Reviewer: ```suggestion
If you are not a developer, there are many other possibilities which do not require programming skills to help QGIS to evolve. Check our [project homepage for more information](https://qgis.org/en/site/getinvolved/index.html).
```
It would be really nice that one day we write about how a person can contribute to such a project, as discussed in a thread few months ago.

Reviewer 2: Agreed...ideas include expanding on documentation (which is great, but a bit dry, "let's make sure we have a sentence about every added feature"; could use lots more explanation of how and why), adding to the new style hub, encouraging people to provide support in the channels mentioned elsewhere....but let's not have this discussion in this thread.
